### PR TITLE
Pub 2366 dvp rtd provider

### PR DESCRIPTION
**Pull request by dv-anatolyg49**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Review by timdp — CHANGES_REQUESTED** [review]
A few more points that don't belong with any specific part of the code:
1. Is there any guidance from Prebid.js around browser support?
2. It's fine that the title of this internal PR was autogenerated from the branch name, I would also like us to prepare the full PR for the public project - i.e., a decent title and body.
3. I focused more on code style than on the actual signals integration, given that the test suite seems to be fairly comprehensive. Can you think of any blind spots? For example, shouldn't we be testing our error handling with an invalid response from pub.json?

File: integrationExamples/gpt/dvpsRtdProvider_example.html [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+          /** https://docs.prebid.org/dev-docs/publisher-api-reference/setConfig.html#Real-Time%20Data%20Modules */
+          auctionDelay: 2000,
+          dataProviders: [
+            {
+              name: "dvps",
+              waitForIt: false,
+              params: {
+                ctx: "prebid-test-publisher",
+                cmp: "prebid-test-tag",
+                signals: ['ids', 'bsc', 'vlp', 'tvp'],
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply integrationExamples/gpt/dvpsRtdProvider_example.html]: The linter is very permissive. I've tried to stick to the formatting of the core (/src) files.

File: modules/dvpsRtdProvider.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
+/**
+ * This module adds DVPS provider to the real time data module
+ * The {@link module:modules/realTimeData} module is required
+ * The module will add custom targetings to ad units
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: Who should I contact?

File: modules/dvpsRtdProvider.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+import { submodule } from '../src/hook.js'
+import { ajaxBuilder } from '../src/ajax.js'
+import * as utils from '../src/utils.js'
+import { getGlobal } from '../src/prebidGlobal.js'
+
+const SUBMODULE_NAME = 'dvps'
+
+const dvpsTargeting = {}
+
+function qsSerialize(obj) {
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: It was a copy-paste. I've made it shorter and prettier as per your suggestion. I have used Object.keys instead of Object.entries because it is more widely supported and it is unclear how it is handled by the build process.

File: modules/dvpsRtdProvider.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  var enc = encodeURIComponent
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: fixed

File: modules/dvpsRtdProvider.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+  const params = {
+    cmp: config.params.cmp,
+    ctx: config.params.ctx,
+    signals: config.params.signals ? config.params.signals.join(',') : undefined,
+    url: (config.params.url || location.href).split('#')[0],
+    adunits: signalsRequestAdUnits,
+  }
+  const queryString = qsSerialize(params)
+  const SIGNALS_URL = 'https://pub.doubleverify.com/signals/pub.json'
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: done

File: modules/dvpsRtdProvider.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  const signalsRequestAdUnits = {}
+  pbjsAdUnits.forEach((adu) => {
+    const sizesStr = getSizesStrFromAdUnit(adu)
+    signalsRequestAdUnits[adu.code] = [sizesStr]
+  })
+
+  const params = {
+    cmp: config.params.cmp,
+    ctx: config.params.ctx,
+    signals: config.params.signals ? config.params.signals.join(',') : undefined,
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js:104]: This is intentional. This is how our API works, it expects a comma separated list of signal names in signals param.

File: modules/dvpsRtdProvider.js [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+/**
+ * Return reqBidsConfigObj.adUnits if it is not empty.
+ * Otherwise return adUnits defined by reqBidsConfigObj.adUnitCodes if it is not empty.
+ * Otherwise return getGlobal().adUnits.
+ */
+function getAdUnits(reqBidsConfigObj) {
+  if (reqBidsConfigObj.adUnits && reqBidsConfigObj.adUnits.length !== 0) {
+    return reqBidsConfigObj.adUnits
+  }
+  const globalAdUnits = getGlobal().adUnits // returns an empty array if called in `init` hook before `setAdUnitspbjs.addAdUnits` call
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: fixed

File: modules/dvpsRtdProvider.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      callback()
+    }
+  })
+}
+
+/**
+ * Return reqBidsConfigObj.adUnits if it is not empty.
+ * Otherwise return adUnits defined by reqBidsConfigObj.adUnitCodes if it is not empty.
+ * Otherwise return getGlobal().adUnits.
+ */
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js]: Removed docs.

File: modules/dvpsRtdProvider.js [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+      return
+    }
+    const targeting = getTargeting(adUnitCode)
+    dataToReturn[adUnitCode] = targeting
+  }, {})
+
+  return dataToReturn
+}
+
+/** @type {RtdSubmodule} */
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.js:195]: Added type annotations

File: modules/dvpsRtdProvider.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
+The purpose of this Real Time Data Provider is to allow publishers to set DVPS targeting on ad units.
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.md]: done

File: modules/dvpsRtdProvider.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+/**
+ * Query stringify object.
+ * @param {Object} obj
+ * @param {string} parentPath
+ * @return {string}
+ */
+function qsSerialize(obj, parentPath) {
+  const params = [];
+  const keys = Object.keys(obj)
+  for (const key of keys) {
```
Comment (timdp) [inline modules/dvpsRtdProvider.js:50]: `for ... of` will (probably) transpile to an ugly IIFE construct to retain scoping. It's probably better to use either a C-style `for` loop or (preferred) an `Array` function. Since you're using the loop to populate the array, I would `return keys.map(...).join('&')`.

File: modules/dvpsRtdProvider.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+**Example:**
+
+To view an example:
+
+- in your cli run:
+
+`gulp serve --modules=dvpsRtdProvider,appnexusBidAdapter,rubiconBidAdapter,trustxBidAdapter`
+
+Your could also use any other bid adapters.
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:42]:
```suggestion
    You can also use any other bid adapters.
```

(Note the four spaces too.)

File: modules/dvpsRtdProvider.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+where:
+- `ctx` (required) - ID associated with the publisher
+- `cmp` (required) - Tag ID
+- `signals` (optional) - Array of signal names. Possible values are: `"ids"`, `"bsc"`, `"vlp"`, `"tvp"`. By default all signals available to the publisher are requested.
+
+**Example:**
+
+To view an example:
+
+- in your cli run:
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:38]:
```suggestion
- in your CLI, run:
```

File: modules/dvpsRtdProvider.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+`gulp serve --modules=dvpsRtdProvider,appnexusBidAdapter,rubiconBidAdapter,trustxBidAdapter`
+
+Your could also use any other bid adapters.
+
+- in your browser, navigate to:
+
+`http://localhost:9999/integrationExamples/gpt/dvpsRtdProvider_example.html`
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:46]: - Does this need to be in backticks?
- Indent with four spaces again.

File: modules/dvpsRtdProvider.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+- `cmp` (required) - Tag ID
+- `signals` (optional) - Array of signal names. Possible values are: `"ids"`, `"bsc"`, `"vlp"`, `"tvp"`. By default all signals available to the publisher are requested.
+
+**Example:**
+
+To view an example:
+
+- in your cli run:
+
+`gulp serve --modules=dvpsRtdProvider,appnexusBidAdapter,rubiconBidAdapter,trustxBidAdapter`
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:40]: I'd make this a `<pre>` block, so with three backticks instead of one. Single backticks are for inline code.

Also, you probably want to indent every line with four spaces, because it's part of the bullet list.

File: modules/dvpsRtdProvider.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        }
+      }
+    ]
+  }
+});
+```
+
+where:
+- `ctx` (required) - ID associated with the publisher
+- `cmp` (required) - Tag ID
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:31]:
```suggestion
- `ctx` (required) - ID associated with the publisher.
- `cmp` (required) - Tag ID.
```

File: modules/dvpsRtdProvider.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
+The purpose of this Real Time Data Provider is to allow publishers to set DoubleVerify Publisher Solution targeting on ad units.
+
+**Usage for Publishers:**
+
+Compile the DVPS Provider into your Prebid build:
+
+`gulp build --modules=dvpsRtdProvider`
```
Comment (timdp) [inline modules/dvpsRtdProvider.md:7]: I'd make this a `<pre>` block, so with three backticks instead of one. Single backticks are for inline code.

File: test/spec/modules/dvpsRtdProvider_spec.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    'BSC': [
+      '84081001'
+    ],
+    'IDS': 0,
+    'VLP': {
+      '/200901344/ca-pub-2905389980697967/DV_HM_IS_TEST_AD_2': {
+        '320x480': '5'
+      }
+    },
+    'gptInsert': true
```
Comment (timdp) [inline test/spec/modules/dvpsRtdProvider_spec.js:64]: Unquoting keys:

```suggestion
      ctx: 'publisher-id',
      cmp: 'tag-id',
      url: 'http://test.localhost:9999/integrationExamples/gpt/dvpRtdProvider_example.html',
      adunits: {
        '/200901344/ca-pub-2905389980697967/DV_HM_IS_TEST_AD_2': [
          '320x480'
        ]
      }
    },
    normalizedUrl: 'healthgrades.com/right-care/coronavirus',
    BSC: [
      '84081001'
    ],
    IDS: 0,
    VLP: {
      '/200901344/ca-pub-2905389980697967/DV_HM_IS_TEST_AD_2': {
        '320x480': '5'
      }
    },
    gptInsert: true
```

**Review by timdp — CHANGES_REQUESTED** [review]
Thanks for the many updates! :+1:

Please don't let my many comments discourage you. I'm deliberately nitpicking more than usual because this will be public. As far as I'm concerned, the PR is looking pretty solid already and there's just some finetuning left.

File: modules/dvpsRtdProvider.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (timdp) [reply modules/dvpsRtdProvider.js]: I made some suggestions in my second review. However, I'll let @pietermees decide whether product or marketing should also be involved.

File: modules/dvpsRtdProvider.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+Add DVPS real time data provider configuration by setting up a Prebid Config:
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.md]: done

File: integrationExamples/gpt/dvpsRtdProvider_example.html [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+        }
+      });
+      pbjs.addAdUnits(adUnits);
+      pbjs.requestBids({
+        bidsBackHandler: sendAdserverRequest,
+        timeout: PREBID_TIMEOUT
+      });
+    });
+
+    function sendAdserverRequest() {
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply integrationExamples/gpt/dvpsRtdProvider_example.html:105]: done

File: integrationExamples/gpt/dvpsRtdProvider_example.html [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    ];
+
+    var pbjs = pbjs || {};
+    pbjs.que = pbjs.que || [];
+  </script>
+
+  <script>
+    var googletag = googletag || {};
+    googletag.cmd = googletag.cmd || [];
+    googletag.cmd.push(function () {
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply integrationExamples/gpt/dvpsRtdProvider_example.html:75]: In the html file arrow functions are not used for maximum compatibility (html is not compiled).
My preference is to use arrow functions for the functions defined and used right away inside of other functions, and use `function` keyword for standalone reusable functions because IMO `function xxx(...) {}` syntax is way more readable than `const xxx = (...) => {}`. Updated the code in the js file accordingly.

File: modules/dvpsRtdProvider.md [hunk 22]
```diff
@@ -0,0 +1,46 @@
+The purpose of this Real Time Data Provider is to allow publishers to set DoubleVerify Publisher Solution targeting on ad units.
+
+**Usage for Publishers:**
+
+Compile the DVPS Provider into your Prebid build:
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply modules/dvpsRtdProvider.md]: done

File: test/spec/modules/dvpsRtdProvider_spec.js [hunk 23]
```diff
@@ -0,0 +1,121 @@
+import { dvpsSubmodule } from 'modules/dvpsRtdProvider.js';
+import * as prebidGlobalLib from 'src/prebidGlobal.js';
+import * as utilsLib from 'src/utils.js';
+import * as ajaxLib from 'src/ajax.js';
+
+describe('DVPS Real time data submodule', function () {
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply test/spec/modules/dvpsRtdProvider_spec.js]: done

File: test/spec/modules/dvpsRtdProvider_spec.js [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+      beforeEach(function () {
+        sandbox = sinon.sandbox.create();
+        sandbox.spy(utilsLib, 'logError');
+      });
+
+      afterEach(function() {
+        sandbox.restore();
+      });
+
+      it('should log error if initializied without parameters', function () {
```
[the remark this answers is not in the document]
Comment (dv-anatolyg49) [reply test/spec/modules/dvpsRtdProvider_spec.js]: done
